File: python/newnanmanager/services/town.py
"""Town management service."""

from typing import Optional

from ..http_client import HttpClient
from ..models import (
    CreateTownRequest,
    Town,
    TownsListData,
    TownDetailResponse,
    UpdateTownRequest,
)


class TownService:
    """城镇管理服务."""

    def __init__(self, http_client: HttpClient) -> None:
        """初始化城镇服务.

        Args:
            http_client: HTTP客户端
        """
        self._http = http_client

    async def list_towns(
        self,
        page: Optional[int] = None,
        page_size: Optional[int] = None,
        search: Optional[str] = None,
        min_level: Optional[int] = None,
        max_level: Optional[int] = None,
    ) -> TownsListData:
        """获取城镇列表.

        Args:
            page: 页码
            page_size: 每页大小
            search: 搜索关键词
            min_level: 最小等级
            max_level: 最大等级

        Returns:
            城镇列表数据
        """
        params: dict[str, int | str] = {
            "page": page or 1,
            "page_size": page_size or 20,
        }

        # 只添加非None的可选参数
        if search is not None:
            params["search"] = search
        if min_level is not None:
            params["min_level"] = min_level
        if max_level is not None:
            params["max_level"] = max_level

        return await self._http.get(
            "/api/v1/towns",
            params=params,
            response_model=TownsListData,
        )

    async def create_town(
        self,
        request: CreateTownRequest,
    ) -> Town:
        """创建城镇.

        Args:
            request: 创建城镇请求

        Returns:
            创建的城镇信息
        """
        return await self._http.post(
            "/api/v1/towns",
            json_data=request,
            response_model=Town,
        )

    async def get_town(self, town_id: int, detail: bool = False) -> TownDetailResponse:
        """获取城镇详情.

        Args:
            town_id: 城镇ID
            detail: 是否返回详细信息（包括镇长和成员）

        Returns:
            城镇详细信息
        """
        params = {"detail": detail} if detail else None
        return await self._http.get(
            f"/api/v1/towns/{town_id}",
            params=params,
            response_model=TownDetailResponse,
        )

    async def update_town(
        self,
        town_id: int,
        request: UpdateTownRequest,
    ) -> Town:
        """更新城镇信息.

        Args:
            town_id: 城镇ID
            request: 更新请求

        Returns:
            更新后的城镇信息
        """
        return await self._http.put(
            f"/api/v1/towns/{town_id}",
            json_data=request,
            response_model=Town,
        )

    async def delete_town(self, town_id: int) -> None:
        """删除城镇.

        Args:
            town_id: 城镇ID
        """
        await self._http.delete(f"/api/v1/towns/{town_id}")

File: typescript/src/apis/town.ts
import { BaseApiFactory } from './base';
import {
  Town,
  CreateTownRequest,
  UpdateTownRequest,
  ListTownsRequest,
  ListTownsResponse,
  ManageTownMemberRequest,
  GetTownMembersRequest,
  TownMembersResponse,
  EmptyResponse,
  ClientConfig
} from '../types';

/**
 * 城镇管理API
 */
export class TownApi extends BaseApiFactory {
  constructor(config: ClientConfig) {
    super(config);
  }

  /**
   * 创建城镇
   */
  createTown = this.createApi<CreateTownRequest, Town>(
    (request) => ({
      method: 'POST',
      url: '/api/v1/towns',
      data: {
        name: request.name,
        level: request.level ?? 0,
        leader_id: request.leaderId,
        qq_group: request.qqGroup,
        description: request.description
      }
    }),
    ({ data }) => data
  );

  /**
   * 获取城镇详情
   */
  getTown = this.createApi<number, Town>(
    (id) => ({
      method: 'GET',
      url: `/api/v1/towns/${id}`
    }),
    ({ data }) => data
  );

  /**
   * 更新城镇信息
   */
  updateTown = this.createApi<{ id: number; request: UpdateTownRequest }, Town>(
    ({ id, request }) => ({
      method: 'PUT',
      url: `/api/v1/towns/${id}`,
      data: {
        name: request.name,
        level: request.level,
        leader_id: request.leaderId,
        qq_group: request.qqGroup,
        description: request.description
      }
    }),
    ({ data }) => data
  );

  /**
   * 删除城镇
   */
  deleteTown = this.createApi<number, EmptyResponse>(
    (id) => ({
      method: 'DELETE',
      url: `/api/v1/towns/${id}`
    }),
    ({ data }) => data
  );

  /**
   * 获取城镇列表
   */
  listTowns = this.createApi<ListTownsRequest, ListTownsResponse>(
    (request) => {
      const params = this.buildQueryParams({
        page: request.page ?? 1,
        page_size: request.pageSize ?? 20,
        search: request.search,
        min_level: request.minLevel,
        max_level: request.maxLevel
      });

      return {
        method: 'GET',
        url: '/api/v1/towns',
        params
      };
    },
    ({ data }) => data
  );

  /**
   * 管理城镇成员
   */
  manageTownMember = this.createApi<{ townId: number; request: ManageTownMemberRequest }, EmptyResponse>(
    ({ townId, request }) => ({
      method: 'POST',
      url: `/api/v1/towns/${townId}/members`,
      data: {
        player_id: request.playerId,
        action: request.action
      }
    }),
    ({ data }) => data
  );

  /**
   * 获取城镇成员列表
   */
  getTownMembers = this.createApi<{ townId: number; request: GetTownMembersRequest }, TownMembersResponse>(
    ({ townId, request }) => {
      const params = this.buildQueryParams({
        page: request.page ?? 1,
        page_size: request.pageSize ?? 50
      });

      return {
        method: 'GET',
        url: `/api/v1/towns/${townId}/members`,
        params
      };
    },
    ({ data }) => data
  );
}
